compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -1547,41 +1330,41 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
 
     private val CFGNode<*>.origin: CFGNode<*> get() = if (this is StubNode) firstPreviousNode else this
 
-    private fun <T : CFGNode<*>> T.computeIncomingFlow(): Pair<FLOW, Int> {
+    private fun CFGNode<*>.mergeIncomingFlow(
```

Comment:
Always forking is a tough decision, but I agree that all this machinery with flowDiff was quite fragile, hardly understandable and sometimes error-prone

I'll check performance and try to compare memory consumption, and I very hope it will be fine, because I actually love this change

Reply:
I have a prototype of the additional improvement I described above, but I'm not satisfied with the amount of null checks it requires. The handling of binary boolean operators looks especially ugly.